File: src/state/reconcile.ts
/**
 * Pending/prior reconciliation (R16a). Run at startup of every CLI command,
 * under the lock for mutating ops, lock-free for read-only ops (R43).
 *
 * Recovery rules:
 *  - If `.prior/` exists: a prior materialization crashed AFTER we moved the
 *    old `.claude/` aside but BEFORE we renamed `.pending/` into place. The
 *    invariant we want is "live `.claude/` is the last successful state".
 *    We restore by renaming `.prior/` back to `.claude/` (overwriting any
 *    partial `.claude/` from a half-completed step c) and discarding `.pending/`.
 *  - If only `.pending/` exists: a prior materialization crashed during step
 *    a (before we touched `.claude/`). The live state is intact. We just
 *    drop `.pending/`.
 *  - If neither exists: no-op.
 *
 * Persist (R22b) uses the same pattern with the per-profile pending/prior
 * dirs. We expose a generic `reconcilePendingPrior` that takes the three
 * paths so the materialize and persist call sites share behavior.
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";

import { atomicRename, pathExists, rmrf } from "./atomic.js";
import { buildPersistPaths, isRootClaudeMdTmpName, type StatePaths } from "./paths.js";

let reconcileCounter = 0;

/**
 * Outcome of a single reconciliation step. Surfaced upward so E5 can print
 * the "(reconciled crashed materialization: restored .claude/ from .prior/)"
 * notice the spec promises. `targetLabel` identifies which target was
 * reconciled — for materialize it's "<root>/.claude/", for persist it's
 * "<profile>/.claude/".
 */
export type ReconcileOutcome =
  | { kind: "none"; targetLabel: string }
  | { kind: "restored-from-prior"; targetLabel: string }
  | { kind: "discarded-pending"; targetLabel: string };

/**
 * Generic reconciliation step. Returns the action taken so the caller can
 * surface a notice. `target`, `pendingDir`, `priorDir` form the protocol's
 * three-way path triple; `targetLabel` is a human-friendly description for
 * messages.
 */
export async function reconcilePendingPrior(
  target: string,
  pendingDir: string,
  priorDir: string,
  targetLabel: string,
): Promise<ReconcileOutcome> {
  const priorExists = await pathExists(priorDir);
  const pendingExists = await pathExists(pendingDir);

  if (priorExists) {
    // The prior dir was rolled aside (step b). Two distinct on-disk states
    // can land us here:
    //   (i)  step c never ran or partially renamed — `.claude/` is missing
    //        or holds half-renamed bytes;
    //   (ii) step c finished but the cleanup of `.prior/` didn't run before
    //        a crash — `.claude/` is the freshly committed content and
    //        `.prior/` is the previous version.
    // Spec invariant ("live `.claude/` is the last successful state") says
    // restore from `.prior/` in both cases — case (ii) is rare and the
    // re-materialize after restore is idempotent.
    //
    // Window-narrowing (multi-reviewer P2, Gemini #3): rename `.claude/` to
    // a per-attempt scratch dir before renaming prior back. Concurrent
    // readers (R43) see EITHER the original `.claude/` OR the restored one,
    // never an empty/missing tree, except for the brief rename window.
    //
    // Order matters (Sonnet review #1): the live bytes get moved to scratch,
    // then prior is renamed back to target, THEN scratch is discarded. If
    // we cleaned up scratch before the priorDir restore and the restore
    // throws, the original live bytes would be permanently lost.
    let scratch: string | null = null;
    if (await pathExists(target)) {
      scratch = `${target}.reconcile-${reconcileCounter++}-${Math.random()
        .toString(36)
        .slice(2, 8)}`;
      try {
        await atomicRename(target, scratch);
      } catch {
        // Rename failed (filesystem doesn't support move-into-existing-dir,
        // or .claude/ is locked). Fall back to rmrf — wider window but the
        // recovery still completes. No scratch to retain.
        await rmrf(target);
        scratch = null;
      }
    }
    await atomicRename(priorDir, target);
    if (scratch) {
      // Restore succeeded; safe to discard the held-aside live bytes. If the
      // cleanup itself fails, the user sees a `.claude.reconcile-*` dir but
      // the live tree is correct.
      await rmrf(scratch).catch(() => undefined);
    }
    if (pendingExists) {
      await rmrf(pendingDir);
    }
    return { kind: "restored-from-prior", targetLabel };
  }

  if (pendingExists) {
    // Step a partially succeeded (or fully succeeded but step b never ran).
    // The live target is untouched; drop .pending/ to clear the slate.
    await rmrf(pendingDir);
    return { kind: "discarded-pending", targetLabel };
  }

  return { kind: "none", targetLabel };
}

/**
 * Reconcile the materialize target (root `.claude/`). Call at the start of
 * any mutating op (use, sync, persist). Cheap on the steady-state path
 * (two stat calls, both ENOENT).
 *
 * cw6/T4: also sweeps any leftover `<projectRoot>/CLAUDE.md.*.tmp` files
 * from a crashed section-splice write (R45 atomic rollback). The live root
 * CLAUDE.md is untouched — only the staging temp file (which would never
 * have been visible to the user since the splice failed before the rename).
 * This is best-effort: if the unlink fails (permissions, etc.) we swallow
 * because the user's CLAUDE.md is fine; the only cost is a bit of debris.
 */
export async function reconcileMaterialize(paths: StatePaths): Promise<ReconcileOutcome> {
  // Ensure metaDir exists so reads on pending/prior (now under `.meta/`)
  // don't surface unrelated ENOTDIR/ENOENT confusingly. Cheap.
  await fs.mkdir(paths.metaDir, { recursive: true });
  // Sweep any leftover projectRoot CLAUDE.md tmps from crashed splices
  // (cw6/T4 R45). Done before the pending/prior reconcile so a totally-
  // crashed materialize cleans both targets in one pass.
  await sweepRootClaudeMdTmps(paths);
  return reconcilePendingPrior(
    paths.claudeDir,
    paths.pendingDir,
    paths.priorDir,
    paths.claudeDir,
  );
}

/**
 * Best-effort cleanup of leftover section-splice temp files for project-root
 * CLAUDE.md (cw6/T4 / R45 crash recovery). Pattern matches the shape of
 * {@link rootClaudeMdTmpPath} via {@link isRootClaudeMdTmpName} so we never
 * sweep an unrelated `.tmp` file the user may have created.
 *
 * Errors are swallowed: the live CLAUDE.md is untouched, and the worst case
 * of a failed sweep is a stale tmp on disk — diagnosable by the operator,
 * not blocking. The reconcile contract is "fix what you can; don't refuse
 * to start the CLI because of debris from an unrelated process".
 */
async function sweepRootClaudeMdTmps(paths: StatePaths): Promise<void> {
  let entries: string[];
  try {
    const dirents = await fs.readdir(paths.projectRoot, { withFileTypes: true });
    entries = dirents.filter((d) => d.isFile()).map((d) => d.name);
  } catch {
    // If projectRoot is unreadable, the rest of the CLI will fail loudly
    // — no point surfacing the failure here.
    return;
  }
  for (const name of entries) {
    if (!isRootClaudeMdTmpName(name)) continue;
    const abs = path.join(paths.projectRoot, name);
    await fs.unlink(abs).catch(() => undefined);
  }
}

/**
 * Reconcile a per-profile persist target. Used during the persist transactional
 * pair recovery. `profileName` selects which profile dir's pending/prior to
 * inspect.
 */
export async function reconcilePersist(
  paths: StatePaths,
  profileName: string,
): Promise<ReconcileOutcome> {
  const persist = buildPersistPaths(paths, profileName);
  // If the profile dir doesn't exist, there's nothing to reconcile.
  if (!(await pathExists(persist.profileDir))) {
    return { kind: "none", targetLabel: persist.targetClaudeDir };
  }
  return reconcilePendingPrior(
    persist.targetClaudeDir,
    persist.pendingDir,
    persist.priorDir,
    persist.targetClaudeDir,
  );
}

File: go/internal/state/reconcile.go
package state

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync/atomic"
)

// ReconcileOutcomeKind enumerates the result of a single pending/prior
// reconciliation step. CLI dispatch (D7) surfaces "(reconciled crashed
// materialization: restored .claude/ from .prior/)" notices from these.
type ReconcileOutcomeKind string

const (
	ReconcileNone               ReconcileOutcomeKind = "none"
	ReconcileRestoredFromPrior  ReconcileOutcomeKind = "restored-from-prior"
	ReconcileDiscardedPending   ReconcileOutcomeKind = "discarded-pending"
)

// ReconcileOutcome carries the per-target result of reconcilation. TargetLabel
// is a human-friendly description used in messages.
type ReconcileOutcome struct {
	Kind        ReconcileOutcomeKind
	TargetLabel string
}

// reconcileCounter monotonically tags scratch dirs so concurrent reconciles
// (defense-in-depth — only one should hold the lock at a time) never collide.
var reconcileCounter atomic.Uint64

// ReconcilePendingPrior performs the generic three-way pending/prior recovery.
//
// Recovery rules:
//   - If .prior/ exists: a prior materialization crashed AFTER step b but
//     potentially BEFORE step c committed. Spec invariant ("live target is
//     the last successful state") says restore from .prior/ in BOTH cases:
//     (i)  step c never ran or partially landed bytes — target is missing
//          or holds half-renamed bytes;
//     (ii) step c finished but .prior/ cleanup didn't run before crash —
//          target is the freshly committed content.
//     Re-materialize after restore is idempotent so case (ii) is safe.
//   - If only .pending/ exists: step a partially succeeded (or step b never
//     ran). Live target is intact — drop .pending/.
//   - If neither exists: no-op.
//
// Window-narrowing: rename target to a per-attempt scratch dir BEFORE
// renaming prior back. Concurrent readers (R43) see EITHER the original
// target OR the restored one, never an empty/missing tree, except for the
// brief rename window. If the rename-aside fails (e.g. FS doesn't support
// move-into-existing-dir, or target is held), fall back to RmRf — wider
// window but recovery still completes.
//
// Order matters: live bytes get moved to scratch, prior is renamed back to
// target, THEN scratch is discarded. Cleaning up scratch BEFORE the prior
// restore would permanently lose the original live bytes if the restore
// fails.
func ReconcilePendingPrior(target, pendingDir, priorDir, targetLabel string) (ReconcileOutcome, error) {
	priorExists, err := PathExists(priorDir)
	if err != nil {
		return ReconcileOutcome{}, err
	}
	pendingExists, err := PathExists(pendingDir)
	if err != nil {
		return ReconcileOutcome{}, err
	}

	if priorExists {
		var scratch string
		targetExists, err := PathExists(target)
		if err != nil {
			return ReconcileOutcome{}, err
		}
		if targetExists {
			scratch = uniqueScratchPath(target)
			if renameErr := AtomicRename(target, scratch); renameErr != nil {
				// Rename-aside failed — fall back to RmRf. Wider window but
				// recovery completes. No scratch to retain.
				if rmErr := RmRf(target); rmErr != nil {
					return ReconcileOutcome{}, fmt.Errorf("reconcile rmrf target after rename-aside failure: %w", rmErr)
				}
				scratch = ""
			}
		}
		if err := AtomicRename(priorDir, target); err != nil {
			return ReconcileOutcome{}, fmt.Errorf("reconcile rename prior back to target: %w", err)
		}
		if scratch != "" {
			// Discard the held-aside live bytes. If cleanup fails, the user
			// sees a `.claude.reconcile-*` dir but the live tree is correct.
			_ = RmRf(scratch)
		}
		if pendingExists {
			if err := RmRf(pendingDir); err != nil {
				return ReconcileOutcome{}, fmt.Errorf("reconcile drop pending after restore: %w", err)
			}
		}
		return ReconcileOutcome{Kind: ReconcileRestoredFromPrior, TargetLabel: targetLabel}, nil
	}

	if pendingExists {
		if err := RmRf(pendingDir); err != nil {
			return ReconcileOutcome{}, fmt.Errorf("reconcile drop pending: %w", err)
		}
		return ReconcileOutcome{Kind: ReconcileDiscardedPending, TargetLabel: targetLabel}, nil
	}

	return ReconcileOutcome{Kind: ReconcileNone, TargetLabel: targetLabel}, nil
}

// uniqueScratchPath returns a peer-of-target path safe for the rename-aside.
// PID + atomic counter + random suffix prevents collisions even under
// crash-spurious leftover scratch dirs from previous runs.
func uniqueScratchPath(target string) string {
	counter := reconcileCounter.Add(1) - 1
	var b [3]byte
	if _, err := rand.Read(b[:]); err != nil {
		// crypto/rand never fails on supported platforms; if it does, fall
		// back to a fixed suffix — the counter alone gives us uniqueness in
		// a single process.
		b = [3]byte{0, 0, 0}
	}
	return fmt.Sprintf("%s.reconcile-%d-%d-%s", target, os.Getpid(), counter, hex.EncodeToString(b[:]))
}

// ReconcileMaterialize reconciles the materialize target (root .claude/).
// Call at the start of any mutating op (use, sync, persist) AFTER the lock is
// acquired (PR23). Cheap on the steady-state path (two stat calls, both
// ENOENT).
//
// Also sweeps any leftover <projectRoot>/CLAUDE.md.*.tmp files from a crashed
// section-splice write (R45 atomic rollback). Best-effort: if unlink fails
// we swallow because the live root CLAUDE.md is fine; only debris remains.
func ReconcileMaterialize(paths StatePaths) (ReconcileOutcome, error) {
	if err := os.MkdirAll(paths.MetaDir, 0o755); err != nil {
		return ReconcileOutcome{}, err
	}
	sweepRootClaudeMdTmps(paths)
	return ReconcilePendingPrior(
		paths.ClaudeDir,
		paths.PendingDir,
		paths.PriorDir,
		paths.ClaudeDir,
	)
}

// sweepRootClaudeMdTmps best-effort cleans up leftover section-splice tmp
// files for project-root CLAUDE.md (R45 crash recovery). Pattern matches the
// shape of RootClaudeMdTmpPath via IsRootClaudeMdTmpName so unrelated user
// .tmp files are never swept.
//
// Errors swallowed: live CLAUDE.md is untouched; the worst case of a failed
// sweep is stale debris on disk, not a blocker. The reconcile contract is
// "fix what you can; don't refuse to start the CLI because of debris from an
// unrelated process".
func sweepRootClaudeMdTmps(paths StatePaths) {
	entries, err := os.ReadDir(paths.ProjectRoot)
	if err != nil {
		// If projectRoot is unreadable, the rest of the CLI fails loudly —
		// no point surfacing the failure here.
		return
	}
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		if !IsRootClaudeMdTmpName(e.Name()) {
			continue
		}
		_ = os.Remove(filepath.Join(paths.ProjectRoot, e.Name()))
	}
}

// ReconcilePersist reconciles a per-profile persist target. Used during the
// persist transactional pair recovery. profileName selects which profile dir
// to inspect; an invalid name returns the validation error so a malicious or
// corrupt active-profile pointer can't escape the profilesDir.
//
// Returns ReconcileNone with no error when the profile dir doesn't exist —
// nothing to reconcile.
func ReconcilePersist(paths StatePaths, profileName string) (ReconcileOutcome, error) {
	persist, err := BuildPersistPaths(paths, profileName)
	if err != nil {
		return ReconcileOutcome{}, err
	}
	exists, err := PathExists(persist.ProfileDir)
	if err != nil {
		return ReconcileOutcome{}, err
	}
	if !exists {
		return ReconcileOutcome{Kind: ReconcileNone, TargetLabel: persist.TargetClaudeDir}, nil
	}
	return ReconcilePendingPrior(
		persist.TargetClaudeDir,
		persist.PendingDir,
		persist.PriorDir,
		persist.TargetClaudeDir,
	)
}

// IsSchemaTooNewError reports whether err is (or wraps) a *SchemaTooNewError.
// Provided for callers that want a one-call test without spelling errors.As.
func IsSchemaTooNewError(err error) bool {
	var target *SchemaTooNewError
	return errors.As(err, &target)
}
